### PR TITLE
Add job group execution limit option

### PR DESCRIPTION
**Pull request by mdegat01**
## Proposed change

Replace `process_lock` decorator job decorator by adding a new `JOB_GROUP` execution limitation option. This option allows a thread to grab the lock for the group and then freely make calls to methods which share this common lock while still locking other callers out. Instead of having public methods using a lock and private methods that route around the lock to allow one method to call another after obtaining the singular lock on the class.

Lock ownership is tracked via a contextvar which is cloned per asyncio task. This way we can easily determine what task has grabbed hold of a job group lock and which haven't.

In terms of one job calling another, each job defines what it has access to irrespective of its parent. So if a job in a job group calls another job in that job group, that works and the chain can continue. If a job calls another job in a different job group or not in a job group at all then the inner job cannot make more calls in the original job group until control returns to the parent.

This is the start of https://github.com/home-assistant/supervisor/issues/1946 . It implements job groups and leverages them. Progress/stage options exist but are not currently used and there is no way to list all running jobs with the API. Those things can come in a subsequent PR.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (which adds functionality to the supervisor)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:
- Link to cli pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast supervisor tests`)
- [ ] Tests have been added to verify that the new code works.

If API endpoints of add-on configuration are added/changed:

- [ ] Documentation added/updated for [developers.home-assistant.io][docs-repository]

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[docs-repository]: https://github.com/home-assistant/developers.home-assistant

### CODE REVIEW

**Review by pvizeli — APPROVED** [review]

File: supervisor/jobs/__init__.py [hunk 1]
```diff
@@ -70,18 +77,28 @@ def ignore_conditions(self, value: list[JobCondition]) -> None:
...
-        if name not in self._jobs:
-            self._jobs[name] = SupervisorJob(self.coresys, name)
+    def new_job(self, name: str, initial_stage: str | None = None) -> SupervisorJob:
+        """Create a new job."""
+        job = SupervisorJob(name, stage=initial_stage)
+        self._jobs[job.uuid] = job
+        return job
+
+    def get_job(self, uuid: UUID | None = None) -> SupervisorJob | None:
+        """Return a job by uuid if it exists. Returns the current job of the thread if uuid omitted."""
```
Comment (agners) [inline supervisor/jobs/__init__.py:87]:
```suggestion
        """Return a job by uuid if it exists. Returns the current job of the asyncio task if uuid omitted."""
```

File: supervisor/exceptions.py [hunk 2]
```diff
@@ -36,6 +36,18 @@ class JobConditionException(JobException):
     """Exception happening for job conditions."""
 
 
+class JobStartException(JobException):
+    """Exception occurred starting a job on thread."""
```
Comment (agners) [inline supervisor/exceptions.py:40]:
```suggestion
    """Exception occurred starting a job in current asyncio task."""
```

File: supervisor/jobs/__init__.py [hunk 3]
```diff
@@ -1,49 +1,56 @@
...
-        self._progress: int = 0
-        self._stage: str | None = None
 
-    @property
-    def progress(self) -> int:
-        """Return the current progress."""
-        return self._progress
+@define
+class SupervisorJob:
+    """Supervisor running job class."""
```
Comment (agners) [inline supervisor/jobs/__init__.py:26]: Can you add a bit of documentation what this class exactly does?

E.g. Tracks a single Supervisor Job. Ensures that only one job is started on a single asyncio task globally, but allows child jobs even on the same asyncio task.

Ideally we should document a bit the big picture how the Job system exactly works, but I don't think that should be here in code.

**Review by agners — APPROVED** [review]
Other than that, LGTM.

File: supervisor/jobs/__init__.py (hunk 3 above)
Comment (mdegat01) [reply supervisor/jobs/__init__.py:30]: I added some more info to the `start` method as that made more sense to me. I also put some comments explaining why the contextvar works. Let me know if that works.

**Pull request merged**